File: app/scrapper/strategies/gobierno.py
import io
import os
import re
from typing import Union
from urllib.parse import urlparse, parse_qs

import boto3
import feedparser
import requests
import scrapy
from requests.adapters import HTTPAdapter
from scrapy.crawler import CrawlerProcess
from urllib3 import Retry

from ..utils import struct_to_timestamp, get_logger
from ..settings import BUCKET_NAME, FEED_URL, MEDIA_PATH, TABLE_NAME

images = {}
logger = get_logger()
dynamodb = boto3.resource("dynamodb")
table = dynamodb.Table(TABLE_NAME)


def get_entry_guid(url) -> str:
    parsed_url = urlparse(url)
    parsed_qs = parse_qs(parsed_url.query)
    return parsed_qs["p"][0]


class GetImagesSpider(scrapy.Spider):
    short_urls = {}
    name = "get_image"

    def __init__(self, entries):
        long_urls = []
        for entry in entries:
            self.short_urls[entry.link] = entry.id
            long_urls.append(entry.link)
        self.start_urls = long_urls
        super().__init__()

    def parse(self, response):
        image_src = response.css("article > div > img::attr(src)").get()
        short_url = self.short_urls[response.url]
        guid = get_entry_guid(short_url)
        images[guid] = image_src


class GobiernoStrategy:
    http = None

    def __init__(self):
        self.init_http_client()
        self.check_news()

    @property
    def last_update(self) -> int:
        query_kwargs = {
            "KeyConditionExpression": "#Source = :source",
            "ExpressionAttributeNames": {"#Source": "Source"},
            "ExpressionAttributeValues": {":source": "gobierno"},
            "ProjectionExpression": "CreatedAt",
            "ScanIndexForward": False,
            "Limit": 1,
        }
        result = table.query(**query_kwargs)
        try:
            last_item = result["Items"][0]
        except IndexError:
            return None
        return last_item["CreatedAt"]

    def init_http_client(self):
        self.http = requests.Session()
        retries = Retry(
            total=3, backoff_factor=0.5, status_forcelist=(500, 502, 503, 504)
        )
        self.http.mount("http://", HTTPAdapter(max_retries=retries))
        self.http.mount("https://", HTTPAdapter(max_retries=retries))

    def check_news(self):
        logger.info("Getting news ...")
        global images
        images = {}
        feed = feedparser.parse(FEED_URL)
        feed_updated = struct_to_timestamp(feed.feed.updated_parsed)
        last_update = self.last_update
        if last_update is not None and last_update == feed_updated:
            logger.info("No news to save.")
            return
        entries = feed.entries
        if last_update is not None:
            # Remove already processed news
            entries = [
                entry
                for entry in entries
                if struct_to_timestamp(entry.published_parsed) > last_update
            ]
        logger.info(f"{len(entries)} news to save.")
        # Get images and save
        self.fetch_images(entries)
        self.put_news(entries)

    def format_entry(self, entry):
        guid = get_entry_guid(entry.id)

        entry = {
            "Id": guid,
            "Title": entry.title,
            "Summary": self.clean_summary(entry.summary),
            "Link": entry.link,
            "Source": "gobierno",
            "CreatedAt": struct_to_timestamp(entry.published_parsed),
        }
        url = images.get(guid)
        if url:
            extension = re.search("\.(jpe?g|png)", url)[1]
            file_name = f"{guid}.{extension}"
            image_path = self.upload_image(url, file_name)
            if image_path:
                entry["Image"] = image_path
        return entry

    def upload_image(self, url: str, file_name: str) -> Union[str, None]:
        if not url:
            return None
        logger.info("Downloading image ...")
        try:
            response = self.http.get(url)
            response.raise_for_status()
        except requests.exceptions.RequestException:
            logger.error("Error while trying to download image.")
            return None
        logger.info("Image downloaded. Saving it into S3 ...")
        s3 = boto3.resource("s3")
        bucket = s3.Bucket(BUCKET_NAME)
        image = io.BytesIO(response.content)
        key = f"/{MEDIA_PATH}/{file_name}"
        bucket.upload_fileobj(image, key, ExtraArgs={"ACL": "public-read",})
        logger.info("Image successfully uploaded to S3.")
        return key

    def put_news(self, news):
        logger.info("Saving news into DynamoDb ...")
        for n in news:
            item = self.format_entry(n)
            table.put_item(Item=item)

    @staticmethod
    def clean_summary(summary: str) -> str:
        return (
            summary.replace("\xa0", " ")
            .replace("[&#8230;]", "...")
            .replace("&#8211;", "-")
            .replace("&#8220;", '"')
            .replace("&#8221;", '"')
        )

    @staticmethod
    def fetch_images(entries):
        logger.info("Getting news images ...")
        process = CrawlerProcess()
        process.crawl(GetImagesSpider, entries)
        process.start()

File: app/scrapper/strategies/gobierno.js
const AWS = require("aws-sdk");
const stream = require("stream");
const axios = require("axios");

const s3 = new AWS.S3();
const dynamodb = new AWS.DynamoDB.DocumentClient();

const dateToTimestamp = (date) => new Date(date).getTime();

const getLatestSourceDate = async () => {
  const { WP_NEWS_URL } = process.env;
  const params = {
    _fields: "date_gmt",
    per_page: 1,
  };
  const response = await axios.get(WP_NEWS_URL, { params });
  const latestNews = response.data[0];
  return dateToTimestamp(latestNews.date_gmt);
};

const getLatestSavedDate = async () => {
  const { TABLE_NAME } = process.env;
  const params = {
    TableName: TABLE_NAME,
    KeyConditionExpression: "#Source = :source",
    ExpressionAttributeNames: { "#Source": "Source" },
    ExpressionAttributeValues: { ":source": "gobierno" },
    ProjectionExpression: "CreatedAt",
    ScanIndexForward: false,
    Limit: 1,
  };
  const { Items } = await dynamodb.query(params).promise();
  if (!Items.length) {
    return null;
  }
  return Items[0].CreatedAt;
};

const getImageInfo = async (href) => {
  const fields = ["guid.rendered", "mime_type"];
  const params = {
    _fields: fields.join(","),
  };
  let response;
  try {
    response = await axios.get(href, { params });
  } catch (error) {
    console.warn(`Failed to get image info (${href}): ${error}`);
    return;
  }
  const { guid, mime_type } = response.data;
  // Convert mime type to extension
  const mimeTypeRegex = new RegExp("(?:.*)/(.*)");
  const extension = mimeTypeRegex.exec(mime_type)[1];
  return { url: guid.rendered, extension };
};

const downloadImage = async (url) => {
  console.log(`Downloading image (${url}) ...`);
  let response;
  try {
    response = await axios.get(encodeURI(url), { responseType: "stream" });
  } catch (error) {
    console.warn(`Failed to download image (${url}): ${error}`);
    return;
  }
  console.log(`Image (${url}) downloaded.`);
  return response;
};

const uploadImage = async (imageResponse, fileName) => {
  console.log(`Uploading image (${fileName}) to S3 ...`);
  const passthrough = new stream.PassThrough();
  const { BUCKET_NAME, MEDIA_PATH } = process.env;
  const Key = `${MEDIA_PATH}/${fileName}`;
  const params = {
    Bucket: BUCKET_NAME,
    Key,
    Body: passthrough,
    ContentType: imageResponse.headers["content-type"],
    ContentLength: imageResponse.headers["content-length"],
    ACL: "public-read",
  };
  imageResponse.data.pipe(passthrough);
  try {
    await s3.upload(params).promise();
  } catch (error) {
    console.warn(`Failed to upload image (${fileName}) to S3: ${error}`);
    return null;
  }
  console.log(`Image (${fileName}) successfully uploaded to S3.`);
  return Key;
};

const processItem = async (item) => {
  const result = {
    Id: item.id,
    Title: item.title.rendered.replace("&#8211;", "–"),
    Summary: item.content.rendered.replace(/<[^>]*>/g, ""),
    Link: item.link,
    Source: "gobierno",
    CreatedAt: dateToTimestamp(item.date_gmt),
  };
  const media = item._links["wp:featuredmedia"];
  if (media.length) {
    // Download news image
    const imageInfo = await getImageInfo(media[0].href);
    if (imageInfo) {
      const { url, extension } = imageInfo;
      // Upload image to S3
      const imageResponse = await downloadImage(url);
      const fileName = `${item.id}.${extension}`;
      const imagePath = await uploadImage(imageResponse, fileName);
      if (imagePath) {
        result.Image = imagePath;
      }
    }
  }
  return result;
};

const fetchNews = async (minDate) => {
  const { WP_NEWS_URL } = process.env;
  const fields = [
    "id",
    "link",
    "title.rendered",
    "content.rendered",
    "date_gmt",
    "_links.wp:featuredmedia.0.href",
  ];
  const params = {
    _fields: fields.join(","),
  };
  console.log("Fetching news ...");
  const response = await axios.get(WP_NEWS_URL, { params });
  const newsPromises = [];
  response.data.forEach((item) => {
    // Skip already saved news
    if (dateToTimestamp(item.date_gmt) > minDate) {
      newsPromises.push(processItem(item));
    }
  });
  console.log(`${newsPromises.length} news to process.`);
  return Promise.all(newsPromises);
};

const saveNews = async (news) => {
  console.log(`Saving ${news.length} news into DynamoDB ...`);
  const { TABLE_NAME } = process.env;
  const putPromises = news.map((Item) => {
    const params = {
      TableName: TABLE_NAME,
      Item,
    };
    return dynamodb
      .put(params)
      .promise()
      .catch((error) => {
        console.warn(
          `Error while trying to save news ${Item.Id} into DynamoDB: ${error}`
        );
      });
  });
  return Promise.all(putPromises).then(() => {
    console.log("News saved into DynamoDB.");
  });
};

exports.start = async () => {
  console.log("Getting news ...");
  const latestSavedDate = await getLatestSavedDate();
  if (latestSavedDate) {
    const latestSourceDate = await getLatestSourceDate();
    if (latestSourceDate <= latestSavedDate) {
      console.log("No news to save.");
      return;
    }
  }
  const news = await fetchNews(latestSavedDate);
  await saveNews(news);

  return {
    statusCode: 200,
    body: JSON.stringify(news),
  };
};
